File: exec/src/main/java/android/databinding/AndroidDataBinding.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.databinding;

import android.databinding.tool.LayoutXmlProcessor;
import android.databinding.tool.util.L;
import android.databinding.tool.util.Preconditions;
import android.databinding.tool.util.StringUtils;
import android.databinding.tool.writer.JavaFileWriter;

import android.databinding.cli.ProcessXmlOptions;
import com.beust.jcommander.JCommander;
import com.beust.jcommander.ParameterException;

import org.apache.commons.io.Charsets;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.IOUtils;
import org.xml.sax.SAXException;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

import javax.swing.filechooser.FileNameExtensionFilter;
import javax.xml.bind.JAXBException;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.XPathExpressionException;

public class AndroidDataBinding {
    public static void main(String[] args)
            throws Throwable {
        ProcessXmlOptions processXmlOptions = new ProcessXmlOptions();
        final JCommander jCommander = new JCommander(processXmlOptions);
        try {
            jCommander.parse(args);
        } catch (ParameterException ex) {
            jCommander.usage();
            System.err.println(ex.getMessage());
            System.exit(1);
        }
        doRun(processXmlOptions);
    }

    public static void doRun(ProcessXmlOptions processXmlOptions) throws Throwable {
        System.out.println(processXmlOptions);
        LayoutXmlProcessor processor = createXmlProcessor(processXmlOptions);
        LayoutXmlProcessor.ResourceInput input = new LayoutXmlProcessor.ResourceInput(
                false,
                processXmlOptions.getResInput(),
                processXmlOptions.getResOutput()
        );
        L.setDebugLog(true);
        processor.processResources(input);
        if (processXmlOptions.shouldZipLayoutInfo()) {
            File outZip = new File(processXmlOptions.getLayoutInfoOutput(),
                    "layout-info.zip");
            FileUtils.forceMkdir(processXmlOptions.getLayoutInfoOutput());
            ZipFileWriter zfw = new ZipFileWriter(outZip);
            processor.writeLayoutInfoFiles(processXmlOptions.getLayoutInfoOutput(), zfw);
            zfw.close();
        } else {
            processor.writeLayoutInfoFiles(processXmlOptions.getLayoutInfoOutput());
        }
    }

    private static LayoutXmlProcessor createXmlProcessor(ProcessXmlOptions processXmlOptions) {
        final ExecFileWriter fileWriter = new ExecFileWriter(processXmlOptions.getResOutput());
        return new LayoutXmlProcessor(
                processXmlOptions.getAppId(),
                fileWriter,
                processXmlOptions.getMinSdk(),
                processXmlOptions.isLibrary(),
                new MyFileLookup()
        );
    }

    static class MyFileLookup implements LayoutXmlProcessor.OriginalFileLookup {

        @Override
        public File getOriginalFileFor(File file) {
            return file;
        }
    }

    static class ZipFileWriter extends JavaFileWriter {
        ZipOutputStream zos;

        public ZipFileWriter(File outZipFile) throws FileNotFoundException {
            FileOutputStream fos;fos = new FileOutputStream(outZipFile);
            zos = new ZipOutputStream(fos);
        }

        @Override
        public void writeToFile(String canonicalName, String contents) {
            throw new RuntimeException("this is only for files not classes");
        }

        @Override
        public void writeToFile(File exactPath, String contents) {
            ZipEntry entry = new ZipEntry(exactPath.getName());
            try {
                zos.putNextEntry(entry);
                zos.write(contents.getBytes(Charsets.UTF_16));
                zos.closeEntry();
            } catch (Throwable t) {
                L.e(t, "cannot write zip file. Filed on %s", exactPath);
            }
        }

        public void close() throws IOException {
            zos.close();
        }
    }

    static class ExecFileWriter extends JavaFileWriter {

        private final File base;

        public ExecFileWriter(File base) {
            this.base = base;
        }

        @SuppressWarnings("Duplicates")
        @Override
        public void writeToFile(String canonicalName, String contents) {
            String asPath = canonicalName.replace('.', '/');
            File f = new File(base, asPath + ".java");
            writeToFile(f, contents);
        }
    }
}

File: exec/src/main/java/android/databinding/AndroidDataBinding.kt
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("MemberVisibilityCanPrivate")

package android.databinding

import android.databinding.cli.GenerateBaseClassesOptions
import android.databinding.cli.ProcessXmlOptions
import android.databinding.cli.ZipUtil
import android.databinding.tool.BaseDataBinder
import android.databinding.tool.DataBindingBuilder
import android.databinding.tool.LayoutXmlProcessor
import android.databinding.tool.store.LayoutInfoInput
import android.databinding.tool.util.L
import android.databinding.tool.writer.JavaFileWriter
import com.beust.jcommander.JCommander
import org.apache.commons.io.Charsets
import org.apache.commons.io.FileUtils
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.nio.file.Files
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

object AndroidDataBinding {
    private const val PROCESS_RESOURCES = "PROCESS"
    private const val GEN_BASE_CLASSES = "GEN_BASE_CLASSES"
    @JvmStatic
    fun main(args: Array<String>) {
        /** Sample run:
         * java -jar android-data-binding-fat.jar \
         *     -classInfoOut test-run/info.zip \
         *     -layoutInfoFiles exec/src/test-data/generateBaseClasses/layout-info-1.zip\
         *     -sourceOut test-run/src.zip \
         *     -package ygt.me
         */
        if (args.isEmpty()) {
            printUsageAndExit()
        }
        try {
            when(args[0]) {
                PROCESS_RESOURCES -> {
                    val processXmlOptions = ProcessXmlOptions()
                    val jCommander = JCommander(processXmlOptions)
                    jCommander.parse(*(args.drop(1).toTypedArray()))
                    processResources(processXmlOptions)
                }
                GEN_BASE_CLASSES -> {
                    val genBaseClassOptions = GenerateBaseClassesOptions()
                    val jCommander = JCommander(genBaseClassOptions)
                    jCommander.parse(*(args.drop(1).toTypedArray()))
                    generateBaseClasses(genBaseClassOptions)
                }
                else -> {
                    printUsageAndExit()
                }
            }
        } catch (e : Exception) {
            e.printStackTrace()
            printUsageAndExit()
        }
    }

    private fun printUsageAndExit() {
        val baseClassArgs = GenerateBaseClassesOptions()
        val baseClass = JCommander(baseClassArgs)

        val processArgs = ProcessXmlOptions()
        val process = JCommander(processArgs)
        println(
                StringBuilder().apply {
                    appendln("Usage: This binary can be used to either process xml resources or generate " +
                            "base classes for data binding.")
                    append("exec.jar $PROCESS_RESOURCES ")
                    process.usage(this)

                    appendln()
                    appendln("exec.jar $GEN_BASE_CLASSES")
                    baseClass.usage(this)
                }.toString()
        )
        process.usage()
        System.exit(1)
    }

    @Deprecated("use processXML",
            ReplaceWith("processResources(processXmlOptions)",
                    "android.databinding.AndroidDataBinding.processResources"))
    @JvmStatic
    fun doRun(processXmlOptions: ProcessXmlOptions) {
        processResources(processXmlOptions)
    }

    @JvmStatic
    fun processResources(processXmlOptions: ProcessXmlOptions) {
        println("options $processXmlOptions")
        val processor = createXmlProcessor(processXmlOptions)
        val input = LayoutXmlProcessor.ResourceInput(
                false,
                processXmlOptions.resInput,
                processXmlOptions.resOutput
        )
        L.setDebugLog(true)
        processor.processResources(input)
        if (processXmlOptions.shouldZipLayoutInfo()) {
            val outZip = File(processXmlOptions.layoutInfoOutput,
                    "layout-info.zip")
            FileUtils.forceMkdir(processXmlOptions.layoutInfoOutput)
            val zfw = ZipFileWriter(outZip)
            processor.writeLayoutInfoFiles(processXmlOptions.layoutInfoOutput, zfw)
            zfw.close()
            L.d("writing info zip to ${outZip.canonicalPath}")
        } else {
            processor.writeLayoutInfoFiles(processXmlOptions.layoutInfoOutput)
        }
    }

    /**
     * Creates data binding base classes and ClassInfo for the given parameters.
     * This should be run after data binding exports the info files from layouts, and before
     * javac runs because it generates sources for javac.
     */
    @JvmStatic
    fun generateBaseClasses(options: GenerateBaseClassesOptions) {
        val classInfoFiles = prepareInput(options.dependencyClassInfoFolders)
        val layoutInfoFolder = prepareInput(arrayListOf(options.layoutInfoFolder))
        val classInfoOutFolder = Files.createTempDirectory("db-class-info-out").toFile()
        val args = LayoutInfoInput.Args(
                outOfDate = emptyList(),
                removed = emptyList(),
                infoFolder = layoutInfoFolder,
                dependencyClassesFolder = classInfoFiles,
                artifactFolder = classInfoOutFolder,
                packageName = options.packageName,
                logFolder = Files.createTempDirectory("db-incremental-log").toFile(),
                incremental = false,
                useAndroidX = options.useAndroidX
        )
        val sourceFileWriter = if (options.zipSourceOutput) {
            ZipFileWriter(options.sourceFileOut)
        } else {
            options.sourceFileOut.mkdirs()
            DataBindingBuilder
                    .GradleFileWriter(options.sourceFileOut.absolutePath)
        }
        BaseDataBinder(LayoutInfoInput(args)).generateAll(sourceFileWriter)
        if (options.zipSourceOutput) {
            (sourceFileWriter as ZipFileWriter).close()
        }
        // we need to zip class info because blaze likes it better
        ZipUtil.zip(classInfoOutFolder, options.classInfoOut)
    }

    /**
     * Creates a tmp folder w/ contents from each file. If the file is a zip, it is unzipped.
     */
    private fun prepareInput(inputs: List<File>): File {
        val outFolder = Files.createTempDirectory("db-class-info").toFile()
        inputs.forEach {
            if (it.isFile) {
                ZipUtil.unzip(it, outFolder)
            } else {
                throw IllegalArgumentException("${it.absolutePath} should've been a file a zip file." +
                        "It is not. Does it exist? ${it.exists()}")
            }
        }
        return outFolder
    }

    private fun createXmlProcessor(processXmlOptions: ProcessXmlOptions): LayoutXmlProcessor {
        val fileWriter = ExecFileWriter(processXmlOptions.resOutput)
        return LayoutXmlProcessor(processXmlOptions.appId, fileWriter, MyFileLookup(),
                processXmlOptions.useAndroidX)
    }

    internal class MyFileLookup : LayoutXmlProcessor.OriginalFileLookup {
        override fun getOriginalFileFor(file: File): File {
            return file
        }
    }

    internal class ZipFileWriter @Throws(FileNotFoundException::class)
    constructor(outZipFile: File) : JavaFileWriter() {
        private val zos: ZipOutputStream

        init {
            val fos = FileOutputStream(outZipFile)
            zos = ZipOutputStream(fos)
        }

        override fun writeToFile(canonicalName: String, contents: String) {
            doWrite(canonicalName.replace(".", "/") + ".java", contents)
        }

        override fun deleteFile(canonicalName: String) {
            throw UnsupportedOperationException("Cannot delete file")
        }

        override fun writeToFile(exactPath: File, contents: String) {
            doWrite(exactPath.name, contents)
        }

        private fun doWrite(entryPath: String, contents: String) {
            val entry = ZipEntry(entryPath)
            try {
                zos.putNextEntry(entry)
                zos.write(contents.toByteArray(Charsets.UTF_8))
                zos.closeEntry()
            } catch (t: Throwable) {
                L.e(t, "cannot write zip file. Filed on %s", entryPath)
            }
        }

        @Throws(IOException::class)
        fun close() {
            zos.close()
        }
    }

    internal class ExecFileWriter(private val base: File) : JavaFileWriter() {

        override fun writeToFile(canonicalName: String, contents: String) {
            val f = toFile(canonicalName)
            writeToFile(f, contents)
        }

        private fun toFile(canonicalName: String): File {
            val asPath = canonicalName.replace('.', '/')
            return File(base, asPath + ".java")
        }

        override fun deleteFile(canonicalName: String) {
            val file = toFile(canonicalName)
            if (file.exists()) {
                file.delete()
            }
        }
    }
}
